protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGateway.java

```diff
@@ -0,0 +1,917 @@
+                        final MqttConnectReturnCode returnCode;
+                        if (authAttempt.failed()) {
+                            log.debug("connection request from client [clientId: {}] rejected, authentication failed",
+                                    endpoint.clientIdentifier(), authAttempt.cause());
+                            returnCode = MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED;
+                        } else {
+                            log.debug(
+                                    "connection request from client [clientId: {}] rejected, connection to backend failed",
+                                    endpoint.clientIdentifier(), result.cause());
+                            returnCode = MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE;
+                        }
+
+                        endpoint.reject(returnCode);
+                    }
+                });
+    }
+
+    private Future<Device> tryAuthenticationWithClientCertificate(final MqttEndpoint endpoint) {
+        if (endpoint.isSsl()) {
+            try {
+                final Certificate[] path = endpoint.sslSession().getPeerCertificates();
+                if (path != null && path.length > 0) {
+                    final Future<Device> authAttempt = authenticateDeviceCertificate(path);
+                    log.debug("authentication with client certificate: {}.",
+                            (authAttempt.succeeded()) ? "succeeded" : "failed");
+                    return authAttempt;
+                }
+            } catch (RuntimeException | SSLPeerUnverifiedException e) {
+                log.debug("could not retrieve client certificate from device endpoint: {}", e.getMessage());
+            }
+        }
+        return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED));
+    }
+
+    private Future<Device> authenticateWithUsernameAndPassword(final MqttEndpoint endpoint) {
+        final MqttAuth auth = endpoint.auth();
+        if (auth == null || auth.getUsername() == null || auth.getPassword() == null) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED,
+                    "device did not provide credentials in CONNECT packet"));
+        } else {
+            final Future<Device> authenticatedDevice = authenticateDevice(auth.getUsername(), auth.getPassword(),
+                    endpoint.clientIdentifier());
+            if (authenticatedDevice == null) {
+                return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+            } else {
+                log.debug("authentication with username/password {}.",
+                        (authenticatedDevice.succeeded()) ? "succeeded" : "failed");
+                return authenticatedDevice;
+            }
+        }
+    }
+
+    private Future<Void> connectGatewayToAmqpAdapter(final Device authenticatedDevice) {
+
```

Comment:
Debug log output would be good here, marking the beginning of the connection creation and its result.